Terraform module for GKE cluster

chaitu6022 opened the pull request:
This PR contains google cloud terraform module for creating google Kubernetes cluster, cluster node pool, virtual private cloud, public subnet, private subnet, Iam member's and a service account.

Attached a terraform Plan Document for the resources it will create.

[plan.txt](https://github.com/k8ssandra/k8ssandra-terraform/files/6250958/plan.txt)

chaitu6022 committed:
Terraform Module for GKE cluster

chrislovecnm reviewed (commented):
Looks great but please add more documentation to the terraform files.

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,100 @@
+# project_k8ssandra
```
```suggestion
# K8ssandra GCP Terraform Example
```

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,100 @@
+# project_k8ssandra
+
+## Standerdization 
```
Love that you have standards - please move this information to a DEVELOPER.md file in the root directory of the project

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,100 @@
+https://github.com/jonbrouse/terraform-style-guide/blob/master/README.md#file-names
+
+
+# folder Structure
```
```suggestion
# Folder Structure
```

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,100 @@
+  -Makefile
+  -README.md
+
+
```
```suggestion
# Prerequisites

GCP Project
```

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,100 @@
+gsutil 4.60
+kubectl 1.17.17
+
+clone this repo, and to initialize the terraform run the following command
```
Lets merge this code, but I need you to file an issue to replace all of this with a `make` command

chaitu6022 replied:
I have made the scripts and tested them locally. for the make command.

chrislovecnm commented on gcp/env/dev.tf:
```diff
@@ -0,0 +1,28 @@
+  name       = var.name
+  region     = var.region
+  project_id = var.project_id
+  //  project_services = var.project_services
```
removed dead code please

chrislovecnm commented on gcp/env/dev.tf:
```diff
@@ -0,0 +1,28 @@
+  //  project_services = var.project_services
+}
+
+# first cluster 
```
Improve comments please

chrislovecnm commented on gcp/modules/vpc/variables.tf:
```diff
@@ -0,0 +1,69 @@
+variable "name" {
```
We need to have permittable instances, please.  Set a variable to turn this on or off.  E2E will use cheaper instances.

chrislovecnm replied:
File an issue to have https://cloud.google.com/compute/docs/instances/preemptible instances

chrislovecnm reviewed (changes requested):
What are we putting in the "tests" folder?  Otherwise some requests for some modifications.

chaitu6022 committed:
k8ssandra google cloud terraform module for GKE cluster

chrislovecnm commented on README.md:
```diff
@@ -1 +1,96 @@
-# k8ssandra-terraform
\ No newline at end of file
+## Standerdization 
```
move to DEVELOPER.md please

chrislovecnm replied:
```suggestion
## Standards 
```

chrislovecnm commented on README.md:
```diff
@@ -1 +1,96 @@
+
+Linting and Verifications standards to be checked with the following commands depending on the file type:
+
+### Terraform
```
We need a "#" header that is the project name, and then an intro to what the project is.

chrislovecnm commented on README.md:
```diff
@@ -1 +1,96 @@
+* [Terraform Google Provider](https://www.terraform.io/docs/providers/google/)
+* [Securely Connecting to VM Instances](https://cloud.google.com/solutions/connecting-securely)
+* [Cloud NAT](https://cloud.google.com/nat/docs/overview)
+* [Kubernetes Engine - Hardening your cluster's security](https://cloud.google.com/kubernetes-engine/docs/how-to/hardening-your-cluster)
```
All the empty files should have a header of what is going into the directory and an note that the work is in progress

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+# K8ssandra GCP Terraform Example
+
+## GCP resources
```
```suggestion
## GCP resources created
```

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+## GCP resources
+* GKE cluster
+* Cluster node pool
+* service account 
```
```suggestion
* Service account 
```

Keep the capitalization consistent, please.

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+
+### Required GCP APIs
+
+The following APIs will be enabled:
```
```suggestion
The following APIs are enabled when the terraform is utilized.
```

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+* Stackdriver Monitoring API
+* IAM Service Account Credentials API
+
+All the tools for the demo are installed. When using Cloud Shell execute the following command in order to setup gcloud cli. When executing this command please setup your region and zone.
```
> All the tools for the demo are installed

I do not understand what you are saying.

> When using Cloud Shell execute the following command in order to setup gcloud cli. When executing this command please setup your region and zone.

Let's not mention Cloud Shell.  Write the instructions so that it works on Linux, and it is tested on Linux.

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+
+### Tools
+
+When not using Cloud Shell, the following tools are required:
```
```suggestion
```

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+
+When not using Cloud Shell, the following tools are required:
+
+* Access to an existing Google Cloud project.
```
Which IAM permissions do then need?

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+
+#### Install kubectl CLI
+
+The kubectl CLI is used to interteract with both Kubernetes Engine and Kubernetes in general. [kubectl CLI Installation instructions](https://cloud.google.com/kubernetes-engine/docs/quickstart) for multiple platforms are available online.
```
Just have them install kubectl using gcloud?

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+The kubectl CLI is used to interteract with both Kubernetes Engine and Kubernetes in general. [kubectl CLI Installation instructions](https://cloud.google.com/kubernetes-engine/docs/quickstart) for multiple platforms are available online.
+
+
+### Authenticate gcloud
```
```suggestion
### GCP Authentication

If you have not run commands using `gcloud` previously, please run the following command.
```

chrislovecnm commented on gcp/README.md:
```diff
@@ -0,0 +1,194 @@
+
+### Authenticate gcloud
+
+Prior to running this demo, ensure you have authenticated your gcloud client by running the following command:
```
```suggestion
```

chrislovecnm commented on gcp/README.md:
````diff
@@ -0,0 +1,194 @@
+
+```console
+# Where the region is us-central1
+gcloud config set compute/region us-central1
````
Should we just have them run `gcloud configure`?

chrislovecnm commented on gcp/README.md:
````diff
@@ -0,0 +1,194 @@
+After the terraform initialization is successful, create your workspace and by using the following command
+
+```
+terraform workspace new <WORKSPACENAME>
````
```suggestion
terraform workspace new <WORKSPACENAME_REPLACEME>
```

chrislovecnm commented on gcp/README.md:
`````diff
@@ -0,0 +1,194 @@
+````
+After the terraform initialization is successful, create your workspace and by using the following command
+
+```
`````
```suggestion
```console
```

chrislovecnm commented on gcp/README.md:
````diff
@@ -0,0 +1,194 @@
+terraform workspace select <WORKSPACENAME>
+```
+
+Update variables.tf and setup your project ID
````
Please provide more instructions here. Assume the user has not idea what variables.tf even is.

chrislovecnm commented on gcp/env/README.md:
````diff
@@ -0,0 +1,90 @@
+    variables.tf (all the variable which required by the terraform modules.)
+
+## Gke cluster example module
+```
````
Need more explanation of what this is.  How do I use this?

chrislovecnm commented on gcp/env/backend.tf:
```diff
@@ -0,0 +1,21 @@
+# copyright 2020 Datastax LLC
```
```suggestion
# Copyright 2021 Datastax LLC
```

chrislovecnm commented on gcp/env/backend.tf:
```diff
@@ -0,0 +1,21 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+terraform {
```
Code comments please

chrislovecnm commented on gcp/env/backend.tf:
```diff
@@ -0,0 +1,21 @@
+terraform {
+  backend "gcs" {
+    #  Delete the backend file for local testing.
+    bucket = "tf-state-files-k8ssandra-testing"
```
Make this a variable please

chrislovecnm replied:
Please file an issue to make this a variable.

chrislovecnm commented on gcp/env/dev.tf:
```diff
@@ -0,0 +1,52 @@
+# copyright 2020 Datastax LLC
```
```suggestion
# Copyright 2021 Datastax LLC
```

chrislovecnm commented on gcp/env/dev.tf:
```diff
@@ -0,0 +1,52 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Module call for creating google compute network
```
```suggestion
# Module used for creating a google compute network.
```

Improve all other code comments.

chrislovecnm commented on gcp/env/outputs.tf:
```diff
@@ -0,0 +1,21 @@
+# copyright 2020 Datastax LLC
```
```suggestion
# Copyright 2021 Datastax LLC
```

chrislovecnm commented on gcp/env/outputs.tf:
```diff
@@ -0,0 +1,21 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+output "endpoint" {
```
Code comments please

chrislovecnm commented on gcp/env/variables.tf:
```diff
@@ -0,0 +1,89 @@
+# copyright 2020 Datastax LLC
```
```suggestion
# Copyright 2021 Datastax LLC
```

chrislovecnm commented on gcp/env/variables.tf:
```diff
@@ -0,0 +1,89 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+variable "name" {
```
Code comments please

chrislovecnm commented on gcp/env/version.tf:
```diff
@@ -0,0 +1,24 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+terraform {
```
Code comments please

chrislovecnm commented on gcp/env/version.tf:
```diff
@@ -0,0 +1,24 @@
+# copyright 2020 Datastax LLC
```
```suggestion
# Copyright 2021 Datastax LLC
```

chrislovecnm commented on gcp/modules/gcs/main.tf:
```diff
@@ -0,0 +1,29 @@
+# copyright 2020 Datastax LLC
```
```suggestion
# Copyright 2021 Datastax LLC
```

chrislovecnm commented on gcp/modules/gcs/main.tf:
```diff
@@ -0,0 +1,29 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+resource "google_storage_bucket" "storage_bucket" {
```
Code comments please

chrislovecnm commented on gcp/modules/gke/main.tf:
```diff
@@ -0,0 +1,70 @@
+# copyright 2020 Datastax LLC
```
```suggestion
# Copyright 2021 Datastax LLC
```

chrislovecnm commented on gcp/modules/gke/main.tf:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  provisioner "local-exec" {
+    command = "sleep 120"
```
If you are doing this to wait for the cluster to come up, this will not work.  We need to run a script to check if the cluster is up.

chrislovecnm commented on gcp/modules/gke/main.tf:
```diff
@@ -0,0 +1,70 @@
+
+    service_account = var.service_account
+    oauth_scopes = [
+      "https://www.googleapis.com/auth/logging.write",
```
I am guessing that the operators will need more permissions, specifically for backups.  We probably will want to use https://cloud.google.com/kubernetes-engine/docs/how-to/hardening-your-cluster#workload_identity

Also, take a look at that document and ensure that we are following any of the steps that are easy.  The private API will not be easy, but we should look at that as an option.

chrislovecnm commented on gcp/modules/iam/main.tf:
```diff
@@ -0,0 +1,37 @@
+# copyright 2020 Datastax LLC
```
header

chrislovecnm reviewed (commented):
Overall it looks good.  Add more documentation and fix all of the file headers.

chaitu6022 committed:
google cloud terraform module documentation.

chaitu6022 committed:
Update README.md

chaitu6022 committed:
Improved terraform module code for GKE cluster

In this commit,

New Firewall rules added.
terraform resources code updated.

chaitu6022 committed:
Google Project Services resource.

- Google Project Services resource added to the IAM module
- oauth scopes to the container cluster node pool.

chrislovecnm reviewed (approved):
Please open two issues to fix two problems that were missed. Otherwise please merge.

Great work
